core/celo_multi_gaspool.go

```diff
@@ -0,0 +1,79 @@
+package core
+
+import "github.com/ethereum/go-ethereum/common"
+
+type FeeCurrency = common.Address
+
+// MultiGasPool tracks the amount of gas available during execution
+// of the transactions in a block per fee currency. The zero value is a pool
+// with zero gas available.
+type MultiGasPool struct {
+	pools       map[FeeCurrency]*GasPool
+	defaultPool *GasPool
+
+	blockGasLimit uint64
+	defaultLimit  float64
+}
+
+type FeeCurrencyLimitMapping = map[FeeCurrency]float64
+
+// NewMultiGasPool creates a multi-fee currency gas pool and a default fallback
+// pool for CELO
+func NewMultiGasPool(
+	blockGasLimit uint64,
+	defaultLimit float64,
+	limitsMapping FeeCurrencyLimitMapping,
+) *MultiGasPool {
+	pools := make(map[FeeCurrency]*GasPool, len(limitsMapping))
+	// A special case for CELO which doesn't have a limit
+	celoPool := new(GasPool).AddGas(blockGasLimit)
+	mgp := &MultiGasPool{
+		pools:         pools,
+		defaultPool:   celoPool,
+		blockGasLimit: blockGasLimit,
+		defaultLimit:  defaultLimit,
+	}
+	for feeCurrency, fraction := range limitsMapping {
+		mgp.getOrInitPool(feeCurrency, &fraction)
+	}
+	return mgp
+}
+
+func (mgp MultiGasPool) getOrInitPool(c FeeCurrency, fraction *float64) *GasPool {
+	if gp, ok := mgp.pools[c]; ok {
+		return gp
+	}
+	if fraction == nil {
+		fraction = &mgp.defaultLimit
+	}
+	gp := new(GasPool).AddGas(
+		uint64(float64(mgp.blockGasLimit) * *fraction),
+	)
+	mgp.pools[c] = gp
+	return gp
+}
+
+// GetPool returns an initialised pool for the given fee currency or
+// initialises and returns a new pool with a default limit.
+// For a `nil` FeeCurrency value, it returns the default pool.
+func (mgp MultiGasPool) GetPool(c *FeeCurrency) *GasPool {
```

Comment:
No, dynamically initializing is not necessary and the assumption did not change. In the `celo-blockchain` implementation the `MultiGasPool` was initialized for every new block and thus the whitelist was up-to-date with changes from the previous block.
The change here is necessary because using the whitelist for initialization has been removed (see below why) and initializing new pools becomes necessary for newly whitelisted currencies that haven't been configured by the user.

Reply:
> In the celo-blockchain implementation the was initialized for every new block and thus the initialized pools from the whitelist were up-to-date with changes from the previous block.

Now the `MultiGasPool` isn't recreated for every block? Was that only necessary because of the fee currencies check?

Reply:
No, it is also created for every new work on a block since the gas-pools need to be reset.

With the statement above I meant that the map of pools in the `celo-blockchain` was also updated "dynamically" for whitelist changes _per block_. We don't need to incorporate "_within-block_" changes to the white-list, so the new feature is not required for something like updating the map of pools dynamically _per-transaction_ within a block.